hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/HASecurityUtils.java

```diff
@@ -0,0 +1,266 @@
+      } else {
+        throw new RuntimeException("Unable to retrieve SCM certificate chain");
+      }
+    } catch (IOException | CertificateException e) {
+      LOG.error("Error while storing SCM signed certificate.", e);
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static void getPrimarySCMSelfSignedCert(CertificateClient client,
+      OzoneConfiguration config, String fetchedSCMId,
+      SCMStorageConfig scmStorageConfig, InetSocketAddress scmAddress)
+      throws IOException {
+
+    PKCS10CertificationRequest csr = generateCSR(client, scmStorageConfig,
+        config, scmAddress, fetchedSCMId);
+
+    CertificateServer rootCAServer =
+        initializeRootCertificateServer(scmStorageConfig.getClusterID(),
+            scmStorageConfig.getScmId(), null);
+
+    rootCAServer.init(new SecurityConfig(config),
+        CertificateServer.CAType.SELF_SIGNED_CA);
+
+    try {
+      X509CertificateHolder certificateHolder = rootCAServer.
+          requestCertificate(csr, KERBEROS_TRUSTED, SCM).get();
+
+      X509CertificateHolder rootCAServerCACertificateHolder =
+          rootCAServer.getCACertificate();
+
+      String pemEncodedCert =
+          CertificateCodec.getPEMEncodedString(certificateHolder);
+
+      String pemEncodedRootCert =
+          CertificateCodec.getPEMEncodedString(rootCAServerCACertificateHolder);
+
+
+      client.storeCertificate(pemEncodedRootCert, true, true);
+      client.storeCertificate(pemEncodedCert, true);
+
+      // Write to the location, so that Default CA server can read the
+      // certificate generated by cert client.
+      CertificateCodec certCodec =
+          new CertificateCodec(new SecurityConfig(config),
+              client.getComponentName());
+
+      certCodec.writeCertificate(certificateHolder);
+
+      // Persist scm cert serial ID.
+      scmStorageConfig.setScmCertSerialId(certificateHolder.getSerialNumber()
+          .toString());
+    } catch (InterruptedException e) {
+      LOG.error("Error while storing SCM signed certificate.", e);
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    } catch (ExecutionException e) {
+      LOG.error("Error while storing SCM signed certificate.", e);
+      throw new RuntimeException(e);
+    } catch (IOException | CertificateException  e) {
+      LOG.error("Error while storing SCM signed certificate.", e);
+      throw new RuntimeException(e);
+    }
+
+  }
+
+  /**
+   * This function creates/initializes a certificate server as needed.
+   * This function is idempotent, so calling this again and again after the
+   * server is initialized is not a problem.
+   *
+   * @param clusterID - Cluster ID
+   * @param scmID     - SCM ID
+   */
+  public static CertificateServer initializeRootCertificateServer(
+      String clusterID, String scmID, SCMCertStore scmCertStore)
+      throws IOException {
+    String subject = "scm-rootca@" + InetAddress.getLocalHost().getHostName();
```

Comment:
Can we keep the root ca subject scm@host to be backward compatible?

Reply:
Done